sn_networking/src/metrics/bad_node.rs

```diff
@@ -0,0 +1,293 @@
+            TimeFrameType::LastSixHours => TimeFrameType::LastDay,
+            TimeFrameType::LastDay => TimeFrameType::LastWeek,
+            TimeFrameType::LastWeek => TimeFrameType::Indefinite,
+            TimeFrameType::Indefinite => TimeFrameType::Indefinite,
+        }
+    }
+}
+
+impl ShunnedCountAcrossTimeFrames {
+    pub fn spawn_background_task(
+        time_based_shunned_count: Family<TimeFrame, Gauge>,
+    ) -> tokio::sync::mpsc::Sender<()> {
+        let (tx, mut rx) = tokio::sync::mpsc::channel(10);
+
+        tokio::spawn(async move {
+            let mut shunned_metrics = ShunnedCountAcrossTimeFrames {
+                metric: time_based_shunned_count,
+                tracked_values: Vec::new(),
+            };
+            let mut update_interval = interval(UPDATE_INTERVAL);
+            update_interval.tick().await;
+
+            loop {
+                tokio::select! {
+                    _ = rx.recv() => {
+                        shunned_metrics.record_shunned_metric();
+
+                    }
+                    _ = update_interval.tick() => {
+                        shunned_metrics.update();
+                    }
+                }
+            }
+        });
+        tx
+    }
+
+    pub fn record_shunned_metric(&mut self) {
+        let now = Instant::now();
+        self.tracked_values.push(TrackedValue {
+            time: now,
+            least_bucket_it_fits_in: TimeFrameType::LastTenMinutes,
+        });
+
+        for variant in TimeFrameType::iter() {
+            let time_frame = TimeFrame {
+                time_frame: variant,
+            };
+            self.metric.get_or_create(&time_frame).inc();
+        }
+    }
+
+    pub fn update(&mut self) {
+        let now = Instant::now();
+        let mut idx_to_remove = Vec::new();
+
+        for (idx, tracked_value) in self.tracked_values.iter_mut().enumerate() {
+            let time_elapsed_since_adding = now.duration_since(tracked_value.time).as_secs();
+
+            if time_elapsed_since_adding > tracked_value.least_bucket_it_fits_in.get_duration_sec()
+            {
+                let time_frame = TimeFrame {
+                    time_frame: tracked_value.least_bucket_it_fits_in,
+                };
+                self.metric.get_or_create(&time_frame).dec();
+
+                let new_time_frame = tracked_value.least_bucket_it_fits_in.next_time_frame();
+                if new_time_frame == TimeFrameType::Indefinite {
+                    idx_to_remove.push(idx);
+                } else {
+                    tracked_value.least_bucket_it_fits_in = new_time_frame;
+                }
+            }
+        }
+        // remove the ones that are now indefinite
+        for idx in idx_to_remove {
+            self.tracked_values.remove(idx);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn update_should_move_to_next_state() -> eyre::Result<()> {
```

Comment:
My only suggestion here would be that these could potentially be several different test cases? However, I could be wrong--I don't know enough about the context.

Reply:
Since the logic is inside a for loop, I think having just a single test would be sufficient 😄